### ISSUE TITLE
how to draw a transparent outline over an image

### ISSUE THREAD

**Issue by Magallo** [dtschump/CImg#214]
Hi all,
I have an image built with several passages. I start drawing it using an external file as background, then I draw over it with some other graphical primitives (draw_rectangle(), draw_rect(), draw_text() etc...). Now, as final passage, I need to draw a sort of 'layer' above it, a sort of outline where the center of the image remains unchanged with only one outlined white circle (draw only the borders and not fill it, i.e. keep the content inside the circle unchanged) and everything outside the circle should be layered with a specific color and a transparency factor. This color should not be opaque, i.e. this layer should not cover and completely overwrite the content of the image, but should be drawn with a transparency. To be more clear, I would like to obtain something similar to a draw_rectangle over all the area, with a color with opaque factor set to 0.5f (or in any case a opaque factor lower than 1.0f). The problem is that if I use the draw_rectangle primitive I got this effect on all the image, instead what I want is keep the center of the image unchanged and draw the transparent color only outside the borders of the cirle. Is that clear enough? I add some images to try to clearify the problem.

As you can see here I have an image example. The external 'dark red' part of the image is obtained starting from the original image and drawing over a blue colored rectangle with a opacity factor of 0.5. The center part of the image is unchanged (i.e. contains the original content without any alteration) and over it there is a white colored outlined circle.

How can I obtain this? Keep in mind that I can't use draw_fill() primitive whereas the content of the image is not filled uniformly but could be an irregular multi colored image.

Please help.

![example](https://user-images.githubusercontent.com/19425083/46484614-62509600-c7fa-11e8-9b06-a0b3739bbc64.png)

**Comment by Magallo** [comment]
Sorry for this further post. More in general I would like to make some questions:
- is there a way to force a specific pixel or a region to be completely transparent?
- can I create an initial image setting all the values to transparent? Now the default is all black and the fill() method accept only a number, no way to set a color (array of unsigned char to set r, b and b).

**Comment by Magallo** [comment]
Sifting the documentation I found a version of draw_image that uses a sprite and a mask, both  as image objects. The problem is that there isn’t a good explanation and a sort of example on how to use it. Doing some experiments I discovered that If I draw a total black circle in the mask, that part of the sprite is drawn completely transparent and this is ok, it’s exactly what I want, but I don’t know what is the right color I have to use for the part outside the circle. I would want a color in the mask that keeps the sprite unchanged but it seems it is applied a sort of inverse formula on the resulting colors that results in some color alteration I do not understand. Unfortunately as I said, there is no explanation that helps me understanding the logic of the mask so I don’t know how to proceed. In any case, is this the right approach to achieve what explained in my initial question? Could someone explain to me how to use this technique possibly with an example? Thanks in advance.
